### PR TITLE
GPSW serialization

### LINKED ISSUE #22: my challenges using the library
I want to use the library to implement an access control using ABE. The process involves communication between two parties. I initiated the cpabe-setup at the receiver end and send the public key through a safe channel to the sender of the message. After receiving the public key and perform the encryption of the message with some sets of access structure and the public key, i sent the resulting ciphertext to the receiver but the decryption never possible, i get segmentation (core dumped) always. what could be the possible problems and how can i correct it?

### PR DESCRIPTION
**Pull request by ThibsG**
Follow up of @tilenmarc's work: https://github.com/tilenmarc/CiFEr/tree/marshalling

This PR adds public key serialization and decryption keys serialization.

Closes #22

### CODE REVIEW

**Comment by tilenmarc** [comment]
Hi! Sorry for a late reply, there were some personal things happening and I couldn't work on this.
I see that you continued my work for GPSW serialization. Nice, thanks!
Before merging I see quite a few problems:
- most importantly, you are basing your work on my branch that is behind the current state of CiFEr. There were commits that changed the code of GPSW scheme and this is not reflected in your serialization. Merging the code I get errors.
- you could put GPSW related code in new files, not `data_ser.*` files.
- documentation needs to be written
Let me know if you are willing to work on the above issues. Alternatively I can take your branch, try to fix it and create a new PR.

**Comment by ThibsG** [comment]
Hello! Alright so I rebased our work with current master, resolving some conflicts indeed.
Hopefully, the 3 points you mentioned are now covered.

File: src/serialization/data_ser.c [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+    // Free the unpacked message
+    octet_ser__free_unpacked(msg, NULL);
+    return 0;
+}
+
+void cfe_FP12_BN254_pack(FP12_BN254 *a, OctetSer *msg) {
+    octet oct;
+    oct.val = cfe_malloc(12 * MODBYTES_256_56 * sizeof(char));
+    FP12_BN254_toOctet(&oct, a);
+
```
Comment (tilenmarc) [inline src/serialization/data_ser.c:261]: Add here:
`oct.max = oct.len;`
to  prevent unconditional jumps.

File: src/serialization/fame_ser.c [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+
+cfe_error cfe_fame_sec_key_read(cfe_fame_sec_key *key, cfe_ser *buf) {
+    CfeFameSecKeySer *msg;
+
+    msg = cfe_fame_sec_key_ser__unpack(NULL, buf->len, buf->ser);
+    if (msg == NULL)
+    {
+        return 1;
+    }
+
```
Comment (tilenmarc) [inline src/serialization/fame_ser.c:121]: Add here:
`cfe_fame_sec_key_init(key);`
to have the key initialized.

File: test/serialization/gpsw_ser.c [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    munit_assert(pk.t.size == pk2.t.size);
+    for (size_t i = 0; i < pk.t.size; i++) {
+        int check = ECP2_BN254_equals(&pk.t.vec[i], &pk2.t.vec[i]);
+        munit_assert(check == 1);
+    }
+
+    cfe_gpsw_free(&gpsw);
+    cfe_vec_free(&sk);
+    cfe_gpsw_pub_key_free(&pk);
+    cfe_ser_free(&buf);
```
Comment (tilenmarc) [inline test/serialization/gpsw_ser.c:45]: Add 
`cfe_gpsw_pub_key_free(&pk2);`
to free the memory.

**Review by tilenmarc — COMMENTED** [review]
Hey,
thanks for your work. Now the structure is nice and things work ok.
Before merging I ran `valgrind` (with command `valgrind --leak-check=full ./cifer_test`) to see if there are some memory leaks and unfortunately there are quite a few. I added some comments to fix some of the leaks but there are more. In particular the comments should fix all the leaks from `data_ser` and `fame_ser`, but there are still some from `gpsw_ser`. If you have time to fix this it would be great. Otherwise I can in the next week spend some time with it. Please let me know. And thank you again for the contribution.

**Comment by ThibsG** [comment]
I fixed the ones you pointed me out, but not the others at the moment.
I will try to fix them as well, stay tuned. If I'm lost, I will reach you anyway :wink:

**Comment by tilenmarc** [comment]
Sounds good, thanks!

File: src/serialization/gpsw_ser.c [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+void cfe_gpsw_key_ser(cfe_gpsw_key *a, cfe_ser *buf) {
+    CfeGpswKeySer msg = CFE_GPSW_KEY_SER__INIT;
+
+    MspSer msp = MSP_SER__INIT;
+    MatSer mat = MAT_SER__INIT;
+    msp.mat = &mat;
+
+    MpzSer *val = cfe_malloc(sizeof(MpzSer) * a->msp.mat.cols * a->msp.mat.rows);
+    cfe_msp_pack(&(a->msp), &msp, val);
```
Comment (ThibsG) [inline src/serialization/gpsw_ser.c:122]: Valgrind is reporting lost blocks that are malloc'ed in this call:
```
==976020== 48 bytes in 1 blocks are definitely lost in loss record 2 of 4
==976020==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==976020==    by 0x4864F2C: cfe_malloc (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x48786B9: cfe_msp_pack (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x487936B: cfe_gpsw_key_ser (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x122A38: test_gpsw_key_ser (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x11F06B: munit_test_runner_exec (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x11FC3F: munit_test_runner_run_test_with_params (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x12010B: munit_test_runner_run_test (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x12028D: munit_test_runner_run_suite (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x1202C9: munit_test_runner_run_suite (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x12093C: munit_suite_main_custom (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x114755: main (in /home/work/CiFEr/build/cifer_test)
==976020== 
==976020== 144 bytes in 1 blocks are definitely lost in loss record 3 of 4
==976020==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==976020==    by 0x4864F2C: cfe_malloc (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x4877CD4: cfe_mat_pack (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x48786A8: cfe_msp_pack (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x487936B: cfe_gpsw_key_ser (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x122A38: test_gpsw_key_ser (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x11F06B: munit_test_runner_exec (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x11FC3F: munit_test_runner_run_test_with_params (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x12010B: munit_test_runner_run_test (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x12028D: munit_test_runner_run_suite (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x1202C9: munit_test_runner_run_suite (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x12093C: munit_suite_main_custom (in /home/work/CiFEr/build/cifer_test)
==976020== 
==976020== 288 bytes in 18 blocks are definitely lost in loss record 4 of 4
==976020==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==976020==    by 0x4864F2C: cfe_malloc (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x4877B14: cfe_mpz_pack (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x4877D47: cfe_mat_pack (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x48786A8: cfe_msp_pack (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x487936B: cfe_gpsw_key_ser (in /home/work/CiFEr/build/libcifer.so)
==976020==    by 0x122A38: test_gpsw_key_ser (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x11F06B: munit_test_runner_exec (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x11FC3F: munit_test_runner_run_test_with_params (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x12010B: munit_test_runner_run_test (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x12028D: munit_test_runner_run_suite (in /home/work/CiFEr/build/cifer_test)
==976020==    by 0x1202C9: munit_test_runner_run_suite (in /home/work/CiFEr/build/cifer_test)
```
but if I'm not wrong I am freeing the memory [below](https://github.com/fentec-project/CiFEr/pull/28/files#diff-c6c0ce6bae55c5587d7fefc7399f7ac194c9e9748403c6cae610b85d98079f32R136).
I am missing something but I looked too much this code to see anything :sweat_smile:

**Review by ThibsG — COMMENTED** [review]
I fixed almost everything Valgrind reported, except for one thing (see comment below)

File: src/serialization/gpsw_ser.c [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+    }
+
+    cfe_gpsw_pub_key_unpack(a, msg);
+    // Free the unpacked message
+    cfe_gpsw_pub_key_ser__free_unpacked(msg, NULL);
+    return 0;
+}
+
+void cfe_gpsw_key_pack(cfe_gpsw_key *a, CfeGpswKeySer *msg, MpzSer *val, OctetSer *octets_val) {
+    cfe_msp_pack(&(a->msp), msg->msp, val);
```
Comment (tilenmarc) [inline src/serialization/gpsw_ser.c:95]: You are packing `msp` again here. So just remove this line (and then probably you do not need `val` in the args) and it should be fine.

**Review by tilenmarc — COMMENTED** [review]
Great work and a huge contribution! I found your bug, fresh eyes help:) After you fix it, I will merge. Thanks again.

File: src/serialization/gpsw_ser.c (hunk 5 above)
Comment (ThibsG) [reply src/serialization/gpsw_ser.c]: Thanks! It worked like a charm :wink:

**Comment by ThibsG** [comment]
Waiting for CI and it should be good now :wink:

**Pull request merged**